src/QldbDriver.ts

```diff
@@ -156,36 +169,95 @@ export class QldbDriver {
+                debug(
+                    `Getting session. Current free session count: ${thisDriver._sessionPool.length}. ` +
+                    `Currently available permit count: ${thisDriver._availablePermits}.`
+                );
+                if (thisDriver._semaphore.tryAcquire()) {
+                    thisDriver._availablePermits--;
+                    try {
+                        let session: QldbSession
+                        if (!startNewSession) {
+                            session = thisDriver._sessionPool.pop();
+                        }
+                        if (startNewSession || session == undefined) {
+                            debug("Creating a new session.");
+                            const communicator: Communicator = 
+                                await Communicator.create(thisDriver._qldbClient, thisDriver._ledgerName);
+                            session = new QldbSession(communicator);
+                        }
+                        return session;
+                    } catch (e) {
+                        // An error when failing to start a new session is always retriable
+                        throw new ExecuteError(e, true, true);
+                    }
+                } else {
+                    throw new SessionPoolEmptyError()
                 }
-                //If it is ISE but not because of transaction expiry, then pick new session and retry the transaction
-                if (isInvalidSessionException(err) && !isTransactionExpiredException(err) ) {
-                    continue;
+            }
+
+        // Release semaphore and if the session is alive return it to the pool and return true
+        const releaseSession: (thisDriver: QldbDriver, session: QldbSession) => boolean = 
+            function(thisDriver: QldbDriver, session: QldbSession): boolean {
+                thisDriver._semaphore.release();
+                thisDriver._availablePermits++;
+                if (session != null && session.isAlive()) {
+                    thisDriver._sessionPool.push(session);
+                    return true;
                 } else {
-                    throw err;
+                    return false;
                 }
-            } finally {
-                if (session != null) {
-                    this._returnSessionToPool(session);
+            }
+        
+        retryConfig = (retryConfig == null) ? this._retryConfig : retryConfig;
+        let session: QldbSession;
+        let startNewSession: boolean = false;
+        for (let retryAttempt: number = 1; true; retryAttempt++) {
+            try {
+                session = null;
+                session = await getSession(this, startNewSession);
+                return await session.executeLambda(transactionLambda);
+            } catch (e) {
+                if (e instanceof ExecuteError) {
+                    if (e.isRetriable) {
```

Comment:
If we introduced this odd spelling here, let's not do that, otherwise I'll just be sad for a bit :)

Reply:
I'll immediately investigate if this can be done (see if we don't have any public facing API functions that say "retriable") after this PR is completed.

Reply:
Thanks - can you create a GitHub issue to track this improvement?